Legacy/bonej/src/main/java/org/bonej/plugins/ParticleDisplay.java

```diff
@@ -0,0 +1,477 @@
+			final int nPoints = ellipsoid.length;
+			// rotate points by eigenvector matrix
+			// and add transformation for centre
+			for (int p = 0; p < nPoints; p++) {
+				final double x = ellipsoid[p][0];
+				final double y = ellipsoid[p][1];
+				final double z = ellipsoid[p][2];
+				ellipsoid[p][0] = x * eV[0][0] + y * eV[0][1] + z * eV[0][2] + centre[0];
+				ellipsoid[p][1] = x * eV[1][0] + y * eV[1][1] + z * eV[1][2] + centre[1];
+				ellipsoid[p][2] = x * eV[2][0] + y * eV[2][1] + z * eV[2][2] + centre[2];
+			}
+
+			final List<Point3f> points = new ArrayList<>();
+			for (final double[] anEllipsoid : ellipsoid) {
+				final Point3f e = new Point3f();
+				e.x = (float) anEllipsoid[0];
+				e.y = (float) anEllipsoid[1];
+				e.z = (float) anEllipsoid[2];
+				points.add(e);
+			}
+			final CustomPointMesh mesh = new CustomPointMesh(points);
+			mesh.setPointSize(1.0f);
+			final float red = 0.0f;
+			final float green = 0.5f;
+			final float blue = 1.0f;
+			final Color3f cColour = new Color3f(red, green, blue);
+			mesh.setColor(cColour);
+			try {
+				univ.addCustomMesh(mesh, "Ellipsoid " + el).setLocked(true);
+			} catch (final NullPointerException npe) {
+				IJ.log("3D Viewer was closed before rendering completed.");
+				return;
+			}
+			// Add some axes
+			displayAxes(univ, centre, eV, radii, 1.0f, "Ellipsoid Axes " + el);
+		}
+	}
+
+	/**
+	 * Draw the particle surfaces in a 3D viewer
+	 *
+	 * @param univ          universe where the centroids are displayed.
+	 * @param surfacePoints points of each particle.
+	 */
+	static void displayParticleSurfaces(final Image3DUniverse univ, final Collection<List<Point3f>> surfacePoints,
+			final int colourMode, final double[] volumes, final double splitValue,
+			final EigenvalueDecomposition[] eigens) {
+		int p = 0;
+		final int nSurfaces = surfacePoints.size();
+		for (final List<Point3f> surfacePoint : surfacePoints) {
+			IJ.showStatus("Rendering surfaces...");
+			IJ.showProgress(p, nSurfaces);
+			if (p > 0 && !surfacePoint.isEmpty()) {
+				Color3f pColour = new Color3f(0, 0, 0);
+				if (colourMode == GRADIENT) {
+					final float red = 1.0f - p / (float) nSurfaces;
+					final float green = 1.0f - red;
+					final float blue = p / (2.0f * nSurfaces);
+					pColour = new Color3f(red, green, blue);
+				} else if (colourMode == SPLIT) {
+					if (volumes[p] > splitValue) {
+						// red if over
+						pColour = new Color3f(1.0f, 0.0f, 0.0f);
+					} else {
+						// yellow if under
+						pColour = new Color3f(1.0f, 1.0f, 0.0f);
+					}
+				} else if (colourMode == ORIENTATION) {
+					pColour = ParticleDisplay.colourFromEigenVector(eigens[p]);
+				}
+				// Add the mesh
+				try {
+					univ.addTriangleMesh(surfacePoint, pColour, "Surface " + p).setLocked(true);
+				} catch (final NullPointerException npe) {
+					IJ.log("3D Viewer was closed before rendering completed.");
+					return;
+				}
+			}
+			p++;
```

Comment:
Could be more readable with
```
		for (int i = 1; i < surfacePoints.size(); i++) {
			IJ.showStatus("Rendering surfaces...");
			IJ.showProgress(i - 1, nSurfaces);
			final List<Point3f> surfacePoint = surfacePoints.get(i);
			if (surfacePoint.isEmpty()) {
				continue;
			}
			final Color3f colour = getColour(i, colourMode, splitValue, volumes, eigens);
			if (!addMesh(univ, surfacePoint, colour, i)) {
				return;
			}
		}
```
For this you'd need to change the parameter type from `Collection` to `List`

Reply:
True & done.